### ISSUE TITLE
{"level":"fatal","msg":"'Web.Kubeconfig.ExtraOpts' expected a map, got 'string'"}

### ISSUE THREAD

**Issue by sense-amid-madness** [fydrah/loginapp#39]
I'm trying to migrate from the old loginapp "objectiflibre/login-app:latest", and cannot get this one here to work. What I am currently getting is

{"level":"fatal","msg":"1 error(s) decoding:\n\n* 'Web.Kubeconfig.ExtraOpts' expected a map, got 'string'","time":"2021-01-02T23:33:48Z"}

in the logs the loginapp Kubernetes pod, same as in the previous issue. I tried to add the suggested snipped to the config, but still the same error.

Any help appreciated, in particular a suggestion for a minimal working example for a config file which can be mounted to /app/config.yaml, using a deployment applied directly with kubectl.

**Loginapp & Dex version**
loginapp: `latest (v3.2.0)`
dex: latest (not relevant yet I think)

**Configuration**

Current loginapp (and eventually dex) configuration (*without secrets*).
I am mounting this in app/config.yaml inside the container (starting the deployment manually, without helm).

```
log:
  level: debug
name: "Kubernetes Authentication"
listen: "https://0.0.0.0:32443"
oidc:
  client:
    id: "loginapp"
    secret: REDACTED
    redirect_url: "https://<server>/callback"
  issuer:
    url: "https://dex.auth/dex"
    rootCA: "/etc/ssl/ca.pem"
  scopes:
    - email
    - groups
  web:
    kubeconfig:
      extraOpts: {}
tls:
  enabled: false
clusters:
  - name: mycluster
    server: https://mycluster:7443
    certificate-authority: |
        -----BEGIN CERTIFICATE-----
       ...
        -----END CERTIFICATE-----
    insecure-skip-tls-verify: false
web:
  mainClientID: "loginapp"
  mainUsernameClaim: "email"
  kubeconfig:
    extraOpts: {}
```

deployment:

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: dex-loginapp
  namespace: auth
  labels:
    app: dex-loginapp
spec:
  replicas: 1
  ...
    spec:
      containers:
      - name: dex-loginapp
        #image: objectiflibre/login-app:latest
        image: quay.io/fydrah/loginapp:latest
        command: ["/loginapp"]
        args: ["serve", "/app/config.yaml"]
        volumeMounts:
        - mountPath: /etc/ssl/ca.pem
          name: dex-ca
          readOnly: true
        - mountPath: /app/config.yaml
          name: loginapp-config
          readOnly: true
      volumes:
      - hostPath:
          path: /cephfs/certs/dex/ssl/ca.pem
          type: File
        name: dex-ca
      - hostPath:
          path: /cephfs/volumes/dex-loginapp/config.yaml
          type: File
        name: loginapp-config
```

**Comment by sense-amid-madness** [comment]
Now I managed to get past the error using the older version 2.5.0, but first was stuck at "no root_ca specified".

With some experimenting, I found that it works with "root_ca" instead of "rootCA" as in the example config from the front page.

Moving back to v3.2.0, I noticed that also here, I can work around the above error by specifying "extra_opts" instead of "extraOpts".

I guess it's my fault and there is something fundamental I do not understand yet - do you have to adapt the names in some way when moving from your config format to the one I use in the example above? I know a bug forum is not the right place to ask, but could you maybe give a hint or a link to documentation which explains why?

Thanks much. Anyway, I have a working configuration now, will report back if anything else creeps up.

**Comment by fydrah** [comment]
Hi @sense-amid-madness ,

This is a known issue reported in #37 .

I just push a fix in v3.2.1 (helm chart v1.0.5).

btw, the change from `root_ca` to `rootCA` has been made in v3.0.0 (check [release note](https://github.com/fydrah/loginapp/releases/tag/v3.0.0) ), all yml keys are now in lower camel case format. This was required for the change to cobra/viper CLI libs, and to allow overrrides from configuration file, environment vars & CLI args.

**Comment by sense-amid-madness** [comment]
Thanks for your help, this explains much of the confusion with the names I had. I thought the fix for the previous issue was already online, so I was wondering why I was still getting it. I am only still confused why I could not work around it by defining the extraOpts manually, but maybe I made a mistake there.

I'll wait for v3.2.1 to show up on quay.io/fydrah and hope this will make everything work.

**Comment by fydrah** [comment]
```
listen: "0.0.0.0:32443" # without scheme
# Application secret. Must be identical across
# all loginapp server replicas ( /!\ this is not the OIDC Client secret)
secret: REDACTED # not present
oidc:
  client:
    redirectURL: "https://<server>/callback" # lower camel case
```
You should compare with https://github.com/fydrah/loginapp#configuration

Yes quay.io take is slow today. You can use master tag if you want to try, it is currently the same as v3.2.1.

**Comment by sense-amid-madness** [comment]
Thanks much. I have everything working now, except one thing. In the old login app, the kubeconfig username was the same as the login ID, which is correct for my cluster. Here, it shows the email. I guess this is related to the configuration field "web.mainUsernameClaim". However, I tried various stuff to get the login id back ("id", "uid", "username", "name"), but it always failed with an error message from the server similar to

```
failed to find a claim matching the main_username_claim 'uid'
internal server error
```

What is the correct configuration option which I need to set here?

**Comment by fydrah** [comment]
You have to check the claim returned by your IdP, for dex they are available at https://YOURDEX/.well-known/openid-configuration, ex:
```
{
[...]
  "claims_supported": [
    "aud",
    "email",
    "email_verified",
    "exp",
    "iat",
    "iss",
    "locale",
    "name",
    "sub"
  ]
}
```

**Comment by sense-amid-madness** [comment]
Thanks, this worked for me and I found the right setting.

Unfortunately, I noticed another thing which is missing compared to the old loginapp, and I do not see an option for this in the config: I do not get a refresh token, only the id token, so that I have to relogin after a day or so. What could be the reason for this?

**Comment by fydrah** [comment]
Hi @sense-amid-madness ,

> Unfortunately, I noticed another thing which is missing compared to the old loginapp, and I do not see an option for this in the config: I do not get a refresh token, only the id token, so that I have to relogin after a day or so. What could be the reason for this?

Refresh token configuration is detailed in documentation: https://github.com/fydrah/loginapp/blob/master/docs/deploy.md#refresh-token

> I also noticed that the value of "web.kubeconfig.defaultContext" is only set into the field "currentContext", but not in the name of the context which is being generated for the user. I am not sure if this is intended, as it requires manual edit of the kubeconfig after it was generated to match the two.

`web.kubeconfig.defaultContext` changes the "default" context name generated only. If you change it, you also must change `clusters[*].contextName` to match the custom name.

**Comment by sense-amid-madness** [comment]
Hi fydrah,

thanks for your help. Based on some documentation I read up on dex, I included "offline_access" in the list of scopes requested by loginapp, and this also resulted in me getting a refresh token.

Is there a difference between the config option "refreshToken: true" and this additional scope, or are both equivalent? Is "offlineAsScope" also another way of achieving the same effect?

**Comment by fydrah** [comment]
Hi @sense-amid-madness ,

`refreshToken: true` is for Helm chart (the variable just setup `offlineAsScope: true` for loginapp), and `offlineAsScope: true` just append the `offline_access` to scopes as you figured out :)

We had an issue #41 for refresh tokens, it is now fixed in [v3.2.2](https://github.com/fydrah/loginapp/releases/tag/v3.2.2)

**Comment by sense-amid-madness** [comment]
Thanks a lot for the help. Since everything is working fine now, I am closing the issue.
